ext/opentelemetry-ext-celery/src/opentelemetry/ext/celery/__init__.py

```diff
@@ -0,0 +1,254 @@
+        # execution, so it **must not** be used to retrieve `request` data.
+        # pylint: disable=no-member
+        task = registry.tasks.get(kwargs.get("sender"))
+        task_id = retrieve_task_id(kwargs)
+
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        # TODO: When the span could be CLIENT?
+        span = self._tracer.start_span(task.name, kind=trace.SpanKind.PRODUCER)
+
+        # apply some attributes here because most of the data is not available
+        span.set_attribute(_TASK_TAG_KEY, _TASK_APPLY_ASYNC)
+        span.set_attribute(_MESSAGE_ID_ATTRIBUTE_NAME, task_id)
+        span.set_attribute(_TASK_NAME_KEY, task.name)
+        set_attributes_from_context(span, kwargs)
+
+        activation = self._tracer.use_span(span, end_on_exit=True)
+        activation.__enter__()
+        attach_span(task, task_id, (span, activation), is_publish=True)
+
+    @staticmethod
+    def _trace_after_publish(*args, **kwargs):
+        # pylint: disable=no-member
+        task = registry.tasks.get(kwargs.get("sender"))
+        task_id = retrieve_task_id(kwargs)
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        # retrieve and finish the Span
+        _, activation = retrieve_span(task, task_id, is_publish=True)
+        if activation is None:
+            logger.warning("no existing span found for task_id=%s", task_id)
+            return
+
+        activation.__exit__(None, None, None)
+        detach_span(task, task_id, is_publish=True)
+
+    @staticmethod
+    def _trace_failure(*args, **kwargs):
+        task = kwargs.get("sender")
+        task_id = kwargs.get("task_id")
+        if task is None or task_id is None:
+            logger.debug(
+                "Unable to extract the Task and the task_id. This version of Celery may not be supported."
+            )
+            return
+
+        # retrieve and pass exception info to activation
+        span, _ = retrieve_span(task, task_id)
+        if span is None:
+            return
+
+        ex = kwargs.get("einfo")
+        if ex is None:
+            return
+        if hasattr(task, "throws") and isinstance(ex.exception, task.throws):
+            return
+
+        span.set_status(
+            Status(
+                canonical_code=StatusCanonicalCode.UNKNOWN,
+                description=str(ex),
+            )
+        )
+
+    @staticmethod
+    def _trace_retry(*args, **kwargs):
+        task = kwargs.get("sender")
+        context = kwargs.get("request")
+        if task is None or context is None:
+            logger.debug(
+                "Unable to extract the Task or the Context. This version of Celery may not be supported."
```

Comment:
might be good to call out what is actually being looked for (the "sender" and "request" keys). Might help someone debug.